### ISSUE TITLE
[Bug] "Undefined symbols:" on Xcode build

### ISSUE THREAD

**Issue by ChurikiTenna** [firebase/firebase-unity-sdk#916]
### [REQUIRED] Please fill in the following fields:

  * Unity editor version: 2022.3.16f1
  * Firebase Unity SDK version: 11.6.0
  * Source you installed the SDK: . unitypackage
  * Problematic Firebase Component: FirebaseAnalytics
  * Other Firebase Components in use: Auth,Firestore,AppCheck
  * Additional SDKs you are using: 
pod install result:
```
Installing AppAuth (1.6.2)
Installing BoringSSL-GRPC (0.0.24)
Installing Firebase (10.15.0)
Installing FirebaseABTesting (10.19.0)
Installing FirebaseAnalytics (10.15.0)
Installing FirebaseAppCheckInterop (10.19.0)
Installing FirebaseAuth (10.15.0)
Installing FirebaseCore (10.15.0)
Installing FirebaseCoreExtension (10.19.0)
Installing FirebaseCoreInternal (10.19.0)
Installing FirebaseFirestore (10.15.0)
Installing FirebaseInstallations (10.19.0)
Installing FirebasePerformance (10.15.0)
Installing FirebaseRemoteConfig (10.19.0)
Installing FirebaseSessions (10.19.0)
Installing FirebaseSharedSwift (10.19.0)
Installing GTMAppAuth (2.0.0)
Installing GTMSessionFetcher (3.2.0)
Installing GoogleAppMeasurement (10.15.0)
Installing GoogleDataTransport (9.3.0)
Installing GoogleSignIn (7.0.0)
Installing GoogleUtilities (7.12.0)
Installing PromisesObjC (2.3.1)
Installing PromisesSwift (2.3.1)
Installing RecaptchaInterop (100.0.0)
Installing abseil (1.20220623.0)
Installing gRPC-C++ (1.50.1)
Installing gRPC-Core (1.50.1)
Installing leveldb-library (1.22.2)
Installing nanopb (2.30909.1)
```
  * Platform you are using the Unity editor on: M1 Mac
  * Platform you are targeting: iOS
  * Scripting Runtime: IL2CPP
  * Pre-built SDK from the [website](https://firebase.google.com/download/unity) or open-source from this repo: _____

### [REQUIRED] Please describe the issue here:

I successfully built Xcode from Build Settings window.
When I try to run the app from Xcode, it fails with error.

```
Undefined symbols:
Linker command failed with exit code 1 (use -v to see invocation)
```
Weird thing is, usually there will be continuing comments after `undefined symbols: `, but it's empty, so there is no way to know what is the actual undefined symbols.

I tried 
cleaning the build folder, 
deleting ModuleCache, 
pod deintegrate, 
deleting Podfile.lock, Pods folder,
pod repo update,
building from the Build Settings again. 
None of these worked.

#### Steps to reproduce:

- Have you been able to reproduce this issue with just the Firebase Unity quickstarts (this GitHub project)?
No. but I intend to try later.

- What's the issue repro rate? (eg 100%, 1/5 etc)
100%

- What happened? How can we make the problem occur?
This is the log:
<img width="524" alt="スクリーンショット 2023-12-27 13 20 59" src="https://github.com/firebase/firebase-unity-sdk/assets/52132649/9c7f9dae-949d-4c61-9306-1ca5b1de5aa0">

I'm willing to share the minimum reproducible project.

**Comment by paulinon** [comment]
Hi @ChurikiTenna,

Are you using Xcode 15? Could you try if adding `-ld_classic` to your other linker flags made any difference?

**Comment by ChurikiTenna** [comment]
Thank you for your reply @paulinon .

Yes. I am using Xcode 15.
However, adding -ld_classic did not make any difference..

<img width="1400" alt="スクリーンショット 2023-12-27 23 23 03" src="https://github.com/firebase/firebase-unity-sdk/assets/52132649/e8864156-a2e6-42f3-8a60-b80bbe18b07e">

**Comment by ChurikiTenna** [comment]
I found detailed error message.

```
ld: Undefined symbols:
_OBJC_CLASS_$_FIRMessaging, referenced from:
       in libFirebaseCppMessaging.a[arm64][4](messaging.mm.o)
  _OBJC_CLASS_$_FIRStorage, referenced from:
       in libFirebaseCppStorage.a[arm64][12](storage_ios.mm.o)
  _OBJC_CLASS_$_FIRStorageMetadata, referenced from:
       in libFirebaseCppStorage.a[arm64][11](metadata_ios.mm.o)
...
```
<img width="578" alt="スクリーンショット 2023-12-28 12 03 20" src="https://github.com/firebase/firebase-unity-sdk/assets/52132649/aa7baa21-40ad-4e9c-a647-a5081b6adbc2">

**Comment by ChurikiTenna** [comment]
Hi @paulinon .
Do you have any clue? Is it not related to the plugin?

**Comment by igloo12** [comment]
I am getting this exact same error and the something happened when I added the -Ld_classic flag

**Comment by providence94** [comment]
same here!

Undefined symbol: absl::lts_20220623::variant_internal::ThrowBadVariantAccess()
Undefined symbol: absl::lts_20220623::optional_internal::throw_bad_optional_access()

these same things!

Due to some Xcode errors I understood that I had to update to 11.8 version
on 20th March around 12PM IST I updated and built the project successfully and pushed a build to textflight
but by 1PM IST the same build stopped working and started throwing this absl errors

and our older working commits has stopped building on Xcode all throwing the same errors :/

**Comment by providence94** [comment]
additionally
the below screenshot shows the error we faced that forced us to update the version to 11.8

![Screenshot 2024-03-22 at 9 06 19 PM](https://github.com/firebase/firebase-unity-sdk/assets/10095266/4ea0a61c-50c2-4ff4-b7fb-3252d7ccb1ae)

and the one below is the error that's thrown by the latest 11.8 version
![Screenshot 2024-03-20 at 21 13 24](https://github.com/firebase/firebase-unity-sdk/assets/10095266/0618450c-5b57-4087-8f23-59395cb6b9b3)

**Comment by providence94** [comment]
from my understanding the problem specifically comes from Firestore
Since after commenting out all our firestore related code in the project(with the library still in the unity project) we tried building the Xcode project and it successfully compiled!

but the moment I use the Firestore namespace anywhere in code
the project fails to compile in Xcode

**Comment by georgemarkosian** [comment]
After some back and forth we figured out the only 2 ways to make our app compile:

1. Leave 11.6.0 and remove the Cloud Storage component from the project
2. Update to 11.8.0 and remove Firestore

None of those are livable as our app relies on both. While discovering it we tried different versions of xCode (15.0.1, 15.1, 15.2, 15.3) and Unity (2022.3.20, 2022.3.22). We also tried removing other plugins with native components (FacebookSDK, AppsFlyer, GoogleSignIn) - no difference. We also tested another app we have that works without Firebase, it's ok.

We believe the issue lies in Firebase and hope it will be resolved soon. It affects our business as we can't produce builds and test the upcoming app update.

**Comment by hsallander** [comment]
We're also having the same issue. Using Unity 2021.3.36, Xcode 15.2, macOS Ventura on an M1 Max. Firebase Unity SDK version 11.8.0 (have tried 11.6.0 as well but with the same results).

Firebase installed using the Package Manager (tgz-files), using the following packages:
* com.google.firebase.analytics
* com.google.firebase.app
* com.google.firebase.auth
* com.google.firebase.crashlytics
* com.google.firebase.firestore
* com.google.firebase.functions
* com.google.firebase.remote-config
* com.google.firebase.storage

Sidenote: Firebase 11.6.0 did work for us when we where on Unity 2021.3.21, but after upgrading to 2021.3.36 we get this error and are unable to build for iOS.

**Comment by providence94** [comment]
@hsallander 
@georgemarkosian 

I have made a separate Bug report for our woes because I understand that this particular bug ticket was opened 3 months ago on a slightly different error scenario (it maybe linked or it may not be idk)
lets all have a cry there

https://github.com/firebase/firebase-unity-sdk/issues/974

**Comment by ChurikiTenna** [comment]
Sorry for being absent.
I ended up fixing the error by creating a new folder, and did the build for iOS in that folder.

Couldn't find what exactly was causing the error, but I guess some kind of caches were doing weird stuff.

**Comment by providence94** [comment]
@ChurikiTenna thanks for getting back. 
Ive seen it in multiple support pages that you should match all the versions of the firebase components you are using otherwise you will end up with some weird errors.

**Comment by shniqq** [comment]
FYI: Faced the same or similar issue, found a fix for our case, see #975 (didnt see this post when starting to write that one)

**Comment by ramibadran** [comment]
Hi

We are facing the same issue as below:

We received an email from Apple requesting builds using the iOS 17 SDK or later (due date mid-April), so we proceeded accordingly. However, after upgrading to MacOS Sonoma 14.4 and Xcode 15.3, we encountered several crashes on devices running iOS 17.3 and 17.4.

To address this issue, we upgraded to Firebase SDK version 11.8.0, previously we were using version 11.6.0. Initially, this resolved the issue, and we were able to successfully build for a couple of days.

However, today, when attempting to create another build, we encountered the attached errors and are unable to proceed further with the build process. We attempted to downgrade to version 11.7.0, but encountered the same error with that version as well.

Would you please give advice or help in getting out of this? it impact our business.

![Screenshot 2024-03-27 at 1 42 25 AM](https://github.com/firebase/firebase-unity-sdk/assets/26247943/112afb58-c4d7-4cf4-8a5a-c568c98932f1)

**Comment by a-maurice** [comment]
Hey all, we haven't been able to reliably reproduce this on our end, which is why it has taken a bit of time to find a solution or workaround.

For people that do have this problem, can you include your Podfile.lock?  Our best guess is that it is somehow getting a version of absl that is incompatible with the library.
Another thing that people can try is manually running `pod repo update` to see if that helps potentially pull in correct pods (assuming that is where the problem lies).

**Comment by hsallander** [comment]
> Hey all, we haven't been able to reliably reproduce this on our end, which is why it has taken a bit of time to find a solution or workaround.
> 
> For people that do have this problem, can you include your Podfile.lock? Our best guess is that it is somehow getting a version of absl that is incompatible with the library. Another thing that people can try is manually running `pod repo update` to see if that helps potentially pull in correct pods (assuming that is where the problem lies).

Thanks for the update @a-maurice . Here's my Podfile.lock file (changed file extension to .txt in order to be able to upload it)
[Podfile.lock.txt](https://github.com/firebase/firebase-unity-sdk/files/14775927/Podfile.lock.txt)

**Comment by providence94** [comment]
@aryansaurav do a pod install verbose and have a look at the versions of all the firebase modules being installed.
also be mindful about other sdks you have installed. undefined symbols can come from anywhere.  
if using FacebookSDK disable Link Frameworks Statically to false.

**Comment by providence94** [comment]
if you fetch the latest version of firebase sdk this issue should be
resolved. if you for some reason have to stick to this version have a look
at the relevant scripts posted somewhere here. i had opened a ticket
regarding this and got it resolved

Best Regards,
Ashwin Shanker G :)

**Comment by a-maurice** [comment]
Sorry this should have been closed with either 11.7.1 (which built against the new internal Pods) or 11.8.0, which is built against the current latest.

If you are still having a problem, it would be better to open a new issue, and include everything, especially the full error logs, and the Podfile.lock file.
